test/membrane/error.spec.js

```diff
@@ -0,0 +1,212 @@
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in blue constructors do not leak', () => {
+    class Foo {
+        constructor() {
+            throw { foo: 'bar' }
+        }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { Foo, expect })
+    evalScript(`
+                try {
+                    new Foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in red functions do not leak', () => {
+    let fn
+    function save(arg) {
+        fn = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
+    evalScript(`
+        function foo() {
+            throw { foo: 'bar' }
+        }
+    `)
+
+    try {
+        fn()
+    } catch(e) {
+        expect(e.foo).toBe(undefined);
+    }
+})
+
+it('non-error objects thrown in red consturctors do not leak', () => {
+    let ctor
+    function save(arg) {
+        ctor = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
+    evalScript(`
+        class Foo {
+            constructor() {
+                throw { foo: 'bar' }
+            }
+        }
+    `)
+
+    try {
+        new ctor()
+    } catch(e) {
+        expect(e.foo).toBe(undefined)
+    }
+})
+
+it('blue extended error objects preserve properties', () => {
+    class CustomError extends Error {
+        constructor(message) {
+            super(message)
+            this.bar = 'baz'
+        }
+
+        get foo() {
+            return 'bar'
+        }
+    }
+
+    class Foo {
+        constructor() {
+            throw new CustomError('foo');
+        }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { Foo, expect })
+    evalScript(`
+                try {
+                    new Foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe('bar')
+                    expect(e.bar).toBe('baz')
+                }
+            `)
+
+})
+
+it('red extended error objects do not leak in blue', () => {
+    let ctor
+    function save(arg) {
+        ctor = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
+    evalScript(`    
+        class CustomError extends Error {
+            constructor(message) {
+                super(message)
+                this.bar = 'baz'
+            }
+
+            get foo() {
+                return 'bar'
+            }
+        }
+
+        class Foo {
+            constructor() {
+                throw new CustomError('foo');
+            }
+        }
+
+        save(Foo)
+    `)
+
+    try {
+        new ctor()
+    } catch(e) {
+        expect(e.bar).toBe(undefined)
```

Comment:
I do not get this... it should be fine... it is a red obj that will be see by the blue side as a blue proxy. This is probably because of the issues that I mentioned above.

Reply:
Like I mentioned in the previous comment red needs to use values, blue needs to use references otherwise this test will fail and it will reveal the leak.